libvast/src/experimental_table_slice_builder.cpp

```diff
@@ -688,281 +332,50 @@ void experimental_table_slice_builder::reserve(
-    },
-    [](const arrow::DurationType& t) -> type {
-      if (t.unit() != arrow::TimeUnit::NANO)
-        die(fmt::format("unhandled Arrow type: Duration[{}]", t.unit()));
-      return type{duration_type{}};
-    },
-    [](const arrow::TimestampType& t) -> type {
-      if (t.unit() != arrow::TimeUnit::NANO)
-        die(fmt::format("unhandled Arrow type: Timestamp[{}]", t.unit()));
-      return type{time_type{}};
-    },
-    [](const arrow::MapType& mt) {
-      return type{map_type{
-        make_vast_type(*mt.key_field()),
-        make_vast_type(*mt.item_field()),
-      }};
-    },
-    [](const arrow::ListType& lt) {
-      const auto& embedded_type = make_vast_type(*lt.value_field());
-      return type{list_type{embedded_type}};
-    },
-    [](const arrow::StructType& st) {
-      std::vector<record_type::field_view> field_types;
-      field_types.reserve(st.num_fields());
-      for (const auto& f : st.fields())
-        field_types.emplace_back(f->name(), make_vast_type(*f));
-      return type{record_type{field_types}};
-    },
-    [](const pattern_extension_type&) -> type {
-      return type{pattern_type{}};
-    },
-    [](const address_extension_type&) -> type {
-      return type{address_type{}};
-    },
-    [](const subnet_extension_type&) -> type {
-      return type{subnet_type{}};
-    },
-    [](const enum_extension_type& et) -> type {
-      return type{et.get_enum_type()};
-    },
-    [](const arrow::FixedSizeBinaryType&) -> type {
-      die("only used in old arrow encoding; remove from sum type access when "
-          "making the experimental encoding the default");
-    },
-  };
-  return caf::visit(f, arrow_type);
-}
-
-VAST_DIAGNOSTIC_PUSH
-VAST_DIAGNOSTIC_IGNORE_DEPRECATED
-/// Enhances a VAST type based on the metadata extracted from Arrow.
-/// Metadata can be attached to both Arrow schema and an Arrow field, and VAST
-/// stores metadata on either of the two, using the exact same structure.
-type enrich_type_with_metadata(
-  type t, const std::shared_ptr<const arrow::KeyValueMetadata>& metadata) {
-  if (!metadata)
-    return t;
-  auto names_and_attrs
-    = std::vector<std::pair<std::string, std::vector<struct type::attribute>>>{};
-  auto name_parser = "VAST:name:" >> parsers::u32 >> parsers::eoi;
-  auto attr_parser = "VAST:attributes:" >> parsers::u32 >> parsers::eoi;
-  for (const auto& [key, value] :
-       detail::zip(metadata->keys(), metadata->values())) {
-    if (!key.starts_with("VAST:"))
-      continue;
-    if (uint32_t index{}; name_parser(key, index)) {
-      if (index >= names_and_attrs.size())
-        names_and_attrs.resize(index + 1);
-      names_and_attrs[index].first = value;
-      continue;
+  auto* nested_builder
+    = &caf::get<type_to_arrow_builder_t<record_type>>(*arrow_builder_);
+  if (num_rows_ == 0 || current_leaf_ == leaves_.end()) {
+    leaves_ = caf::get<record_type>(layout()).leaves();
+    current_leaf_ = leaves_.begin();
```

Comment:
Why is this not in the constructor?

Reply:
We need to do this again after adding a row.